### ISSUE TITLE
[GoldSource] GL_MAXTEXTURES crash.

### ISSUE THREAD

**Issue by basuritashka** [ValveSoftware/halflife#2234]
Sometimes when you are on a server with a lot of textures (I guess this is the reason based on the error code) and you keep switching between servers, your client will crash randomly. I can't really tell how to reproduce it so if someone has extra info would be amazing.

**Comment by basuritashka** [comment]
This is a common issue on a lot of mod servers when switching back and forth on them, if someone got the original output on when this crash occurs, would help a lot.

**Comment by SamVanheer** [comment]
There are a maximum of 4800 textures that can be loaded. Normally textures are unloaded when the client starts preparing to load a new map. It looks like this part works fine.

The part that doesn't is the code that reuses texture entries that were freed:
```
while ( HIBYTE(v12->servercount) & 0x80 )
{
  if ( !v9 )
    v9 = v12;
  ++v13;
  ++v12;
  if ( v13 >= v11 )
    goto LABEL_10;
}

//Other stuff that is ignored

LABEL_10:
if ( !v9 )
{
    v9 = &gltextures[numgltextures++];
    if ( numgltextures >= 4800 )
      Sys_Error("Texture Overflow: MAX_GLTEXTURES");
}
```

This is from `GL_LoadTexture2`. It translates to this:
```
gltexture_t* pSlot = nullptr;

for( int i = 0; i < numgltextures; ++i )
{
    if( gltextures[ i ].servercount >= 0 )
        break;

    if( !pSlot  )
        pSlot  = &gltextures[ i ];
}

//Other stuff that is ignored

if ( !pSlot )
{
    pSlot = &gltextures[numgltextures++];
    if ( numgltextures >= 4800 )
      Sys_Error("Texture Overflow: MAX_GLTEXTURES");
}
```

This code is supposed to find the first free entry (servercount < 0), but it instead stops searching when it encounters a single texture in use.

As a result `GL_LoadTexture2` will almost never reuse freed entries and will keep allocating new entries until it runs out, resulting in the fatal error.

To fix this, the loop needs to be changed to this:
```
for( int i = 0; i < numgltextures; ++i )
{
    if( gltextures[ i ].servercount >= 0 )
        continue;

    if( !pSlot  )
        pSlot  = &gltextures[ i ];
}
```

Now it will skip entries in use and correctly reuse free entries.

@mikela-valve This should be fairly simple to fix and test.

**Comment by basuritashka** [comment]
Thanks @SamVanheer appreciated your work.

@kisak-valve fix proposed? Thanks.

**Comment by SamVanheer** [comment]
Sorry, i got this wrong. There's an outer loop that ensures it iterates over each texture.

I did find what i think is the problem: in `GL_LoadTexture2` the `servercount` variable is set depending on the texture type:
```
if ( textureType == GLT_WORLD )
{
  slot->servercount = gHostSpawnCount;
}
else
{
  slot->servercount = 0;
}
```

And in `GL_UnloadTextures`:
```
if ( pTexture->servercount > 0 && pTexture->servercount != gHostSpawnCount )
```

All textures other than the world (map textures) are never unloaded. An easy fix is to always use `gHostSpawnCount` and let all textures unload, but then there's the risk of unloading textures that are only loaded once (e.g. `Draw_PicFromWad`).

`GLT_SYSTEM` and `GLT_DECAL` texture types definitely shouldn't be unloaded, it looks like all others are resources that are loaded once per map. Maybe allowing all of these to be unloaded is the safest bet:
* GLT_STUDIO
* GLT_WORLD
* GLT_SPRITE

**Comment by mikela-valve** [comment]
I'm going to delay this until after the next release as I'd like it to be in a beta that has fewer changes in it to make it easier to notice if any issues arise from unloading these other texture types.

**Comment by basuritashka** [comment]
@mikela-valve Now that we got a clean beta and the latest update got released officially would like to take a look at this issue because is a very common one on modded servers.

**Comment by PM32** [comment]
Any news about this? It's a bit annoying nowadays

**Comment by zorkenJB** [comment]
@mikela-valve can you fix this?

**Comment by TheDoctor0** [comment]
I would also gladly see this fix applied as it is a common issue nowadays.

**Comment by PM32** [comment]
Nothing? @mikela-valve it is a frequent error this days, will this ever be considerated on a future release?

**Comment by metita** [comment]
@mikela-valve Sorry for bothering you with this one, is there a chance for this issue to be somehow added into the beta now that there are almost 0 issues with it? It is a annoying bug that is pretty common nowadays.

**Comment by Roccoxx** [comment]
considerating we're getting updates for the game, can you please take a look at this issue? @mikela-valve  thanks in advance

**Comment by Javierj12** [comment]
This is very annoying, can you take a look at it please? @mikela-valve

**Comment by Solokiller** [comment]
You don't all have to get his attention, one person calling for him is more than enough. And the issue is already assigned to the next release milestone so it's a priority issue anyway.

**Comment by AgressionTerror** [comment]
When will it be? after 5 years? or will not at all. mikela-valve

**Comment by PM32** [comment]
This is a high priority issue which is ruining actual gameplay, and also an excuse for players to use pirate software (to fix it) because most servers are modded nowadays with lots of resources, bringing a high chance to suffer this crash - @mikela-valve may I ask *again*, is there a chance to get this fixed?

**Comment by cyb3rm4n** [comment]
Agreed with @PM32, got this error on most of servers (i have my own modded server and my client suffers with this error)

**Comment by zorkenJB** [comment]
@mikela-valve can you please fix it? is not that hard
    On Saturday, November 14, 2020, 01:42:51 AM GMT+2, Alexander Zhukov <notifications@github.com> wrote:  
 
 


Agreed with @PM32, got this error on most of servers (i have my own modded server and my client suffers with this error)

**Comment by Maxi605** [comment]
> @mikela-valve can you please fix it? is not that hard On Saturday, November 14, 2020, 01:42:51 AM GMT+2, Alexander Zhukov <notifications@github.com> wrote: Agreed with @PM32, got this error on most of servers (i have my own modded server and my client suffers with this error) — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

Is not about difficulty, he doesn't have time to work on goldsrc.

**Comment by orinasa** [comment]
> > @mikela-valve can you please fix it? is not that hard On Saturday, November 14, 2020, 01:42:51 AM GMT+2, Alexander Zhukov [notifications@github.com](mailto:notifications@github.com) wrote: Agreed with @PM32, got this error on most of servers (i have my own modded server and my client suffers with this error) — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> Is not about difficulty, he doesn't have time to work on goldsrc.
15 minutes work. and you say he doesn't have time to work? Lol....

**Comment by Solokiller** [comment]
Supporting a 22 year old game is about as low priority as it gets, and chances are the source code isn't even accessible remotely so they can't work on it until the pandemic is under control. All you can do is be patient.

**Comment by s1lentq** [comment]
> Sorry, i got this wrong. There's an outer loop that ensures it iterates over each texture.
> 
> I did find what i think is the problem: in `GL_LoadTexture2` the `servercount` variable is set depending on the texture type:
> 
> ```
> if ( textureType == GLT_WORLD )
> {
>   slot->servercount = gHostSpawnCount;
> }
> else
> {
>   slot->servercount = 0;
> }
> ```
> 
> And in `GL_UnloadTextures`:
> 
> ```
> if ( pTexture->servercount > 0 && pTexture->servercount != gHostSpawnCount )
> ```
> 
> All textures other than the world (map textures) are never unloaded. An easy fix is to always use `gHostSpawnCount` and let all textures unload, but then there's the risk of unloading textures that are only loaded once (e.g. `Draw_PicFromWad`).
> 
> `GLT_SYSTEM` and `GLT_DECAL` texture types definitely shouldn't be unloaded, it looks like all others are resources that are loaded once per map. Maybe allowing all of these to be unloaded is the safest bet:
> 
> * GLT_STUDIO
> * GLT_WORLD
> * GLT_SPRITE
> 
> It looks like hud sprites are always `GLT_HUDSPRITE` so they should be safe.

@Solokiller IMHO, it looks like the issue is that the `numgltexture` global counter never decreases, and also textures potentially can be unloaded from the middle of the `gltexture_t gltextures[4800]` without shifting elements

**Comment by Solokiller** [comment]
True, it doesn't seem to ever decrementthat variable. However it does reuse texture slots that were unloaded so the error should only occur if all 4800 slots are in use at the same time. If the error occurs after a certain amount of time rather than when loading a particular map or other assets then it isn't clearing those slots correctly.

**Comment by MatiasEsf** [comment]
three years later, without news, we also provide the code with the solution ... they just have to implement and test it.

**Comment by lexzor** [comment]
@mikela-valve up?

**Comment by huihuangex** [comment]
anyone fix it？

**Comment by anzz1** [comment]
@SamVanheer I tested your proposed fix, unloading textures GLT_STUDIO, GLT_WORLD, GLT_SPRITE
by changing
`if ( textureType == GLT_WORLD )`
to
`if ( textureType >= GLT_STUDIO )`
but it ends up unloading textures that are never reloaded, e.g. viewmodels for default weapons.
The error goes away, but you end up with untextured models instead (models are solid grey color).

**Comment by SpaceArbalest** [comment]
Hello everyone, I found that a developer on github from China has completely fixed the 'Texture overflow: GL_MAXTEXTURES crash' issue in the GoldSrc engine version of Counter-Strike. Here’s what he wrote in the changelog: 'Removes the MAX_GLTEXTURES restriction and optimizes the texture management mechanism, which also fixes the rendering issues after changing map multiple times'

https://github.com/StellarRach/Counter-Strike-Augmented-Edition

One more thing I need to tell you is that this Counter-Strike augmented edition is developed based on Counter-Strike 1.6 build 3266. At that time, the CS1.6 version 3266 did not support mouse raw input, so you need to add a shortcut for the main program of Counter-Strike augmented edition, along with three launch options. These three options are: -noforcemaccel, -noforcemparms, and -noforcemspd (note that there should be a space before the "-" in the first option). Additionally, you need to input the following command in the console: fps_max 99.5 (if you prefer to play Counter-Strike augmented edition at 60 fps, replace 99.5 with 60.5, and other fps values can be adjusted accordingly). This will resolve the issue where "Mouse Acceleration" cannot be completely removed in pirated GoldSrc version Counter-Strike under Windows 11 or Windows 10.

**Comment by SpaceArbalest** [comment]
> anyone fix it？

哥们，我知道你是中文使用者，我也是中文使用者。我需要告诉你一件事情，那就是GitHub上有个中国开发者彻底解决了金源引擎版《Counter-Strike》在部分地图中或在使用某些特效多的武器时游戏必然会崩溃的问题。此外，这位开发者的版本并不会像百度贴吧里的csoldjb吧里发布的CSSME那样，虽然解决了纹理上限问题但却会出现贴图错乱的BUG。这个Counter-Strike augmented edition我已经测试了数月，完全不会出现因为纹理溢出而崩溃的问题，也不会出现贴图错乱，唯一美中不足的就是这个Counter-Strike augmented edition并不像韩国Nexon公司二次开发的《Counter-Strike Online》或《Counter-Strike Nexon》那样支持Hardware Acceleration，否则Counter-Strike augmented edition在许多人使用武器特效多的武器时也能轻松保持99 fps了。

**Comment by SpaceArbalest** [comment]
> anyone fix it？

还有一点，我需要告诉你，这个Counter-Strike augmented edition是基于Counter-Strike 1.6的build 3266版本开发，当时3266版本的CS1.6并不支持mouse的raw input，所以你需要为Counter-Strike augmented edition的主程序添加一个shortcut，同时添加三个启动项，这三个启动项分别是，-noforcemaccel -noforcemparms -noforcemspd（第一个-noforcemaccel的“-”的前面需要加空格），同时你需要在console输入：fps_max 99.5（如果你喜欢玩60 fps的Counter-Strike augmented edition则将99.5改为60.5，其他的fps同样以此类推），这样便可以解决盗版的goldsrc版Counter-Strike在Windows 11或Windows 10下，“Mouse Acceleration”无法彻底移除的问题。
